### PR TITLE
[CPU] Enable vector.interleave lowering

### PR DESCRIPTION
**Pull request by dcaballe**
This PR adds patterns to properly lower the `vector.interleave` ops that are generated when emulating i4 types after the last integrate.

### CODE REVIEW

**Comment by dcaballe** [comment]
The last integrate is breaking the compilation of some internal models. This PR should fix it.

**Comment by hanhanW** [comment]
Are we able to move them to `LLVMCPUVector*` passes?

**Comment by dcaballe** [comment]
Unfortunately, not. These ops are generated by the vector emulation, which happens after the vector lowering in the pipeline. The patterns are mostly preparing to ops to be lowered to LLVM.

**Comment by hanhanW** [comment]
I see... In this case, can you add some e2e tests to https://github.com/openxla/iree/tree/main/tests/e2e/linalg ? Landing a PR without tests is scaring to me, especially when it seems like we're fixing something.

**Comment by dcaballe** [comment]
I thought this fell into the "pipeline" changes that we don't add tests for. I can try to extract a test from the model but it's a case where an i4 bitcast is generated and gets trapped between two tensor.insert ops and the bubbling up is needed to move the bitcast out of them. I don't know which high level ops lead to that situation. Let me take a look but it might be easier to add a before/after test with the specific IR.

**Comment by hanhanW** [comment]
> I thought this fell into the "pipeline" changes that we don't add tests for. I can try to extract a test from the model but it's a case where an i4 bitcast is generated and gets trapped between two tensor.insert ops and the bubbling up is needed to move the bitcast out of them. I don't know which high level ops lead to that situation. Let me take a look but it might be easier to add a before/after test with the specific IR.

Cool, that's even better! I thought that it is hard, so I suggested the e2e tests.

**Comment by dcaballe** [comment]
Done! but the test won't work until we don't cherry-pick the missing commit from LLVM.

**Comment by hanhanW** [comment]
> Done! but the test won't work until we don't cherry-pick the missing commit from LLVM.

It's merged in https://github.com/openxla/iree/pull/16609 I started a new integrate this afternoon, and I just landed it.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir [hunk 1]
```diff
@@ -160,3 +160,45 @@ module attributes {
...
+    return
+  }
+}
+
+// Make sure we can lowering multi-dimensional `vector.interleave` and its
+// corresponding multi-dimensional `vector.bitcast`.
+
+// CHECK-LABEL: llvm.func @interleave_and_bitcast_lowering()
+//   CHECK-NOT:   vector.bitcast %{{.*}} : vector<4x4xi4> to vector<4x2xi8>
+//   CHECK-NOT:   vector.interleave
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir:83]: Does the `CHECK-NOT` make sense? This is dialect conversion, so I thought they should all be converted to llvmir dialect?

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir:83]: This pass is doing quite a few things, including apply patterns greedily. The test will fail with an unrealized conversion that can't be removed but the CHECK will still work:

```
iree3/compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir:202:15: error: CHECK-NOT: excluded string found in input                                                                                                                                                                       
// CHECK-NOT: vector.bitcast %{{.*}} : vector<4x4xi4> to vector<4x2xi8> 
```

**Comment by dcaballe** [comment]
Not sure what happened but I submitted a merge + new commit to the branch (I can see it's there) but I can't see it here :(

**Comment by dcaballe** [comment]
Oh, now :)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir:83]: I see, thanks for the details!

**Review by hanhanW — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir [hunk 2]
```diff
@@ -40,3 +40,164 @@ module {
...
+//      CHECK:   %[[INSERT_ARG2:.+]] = llvm.insertvalue %[[Cf64]], %[[INSERT_ARG1]][3]
+//      CHECK:   %[[INSERT_ARG3:.+]] = llvm.insertvalue %[[DATA]], %[[INSERT_ARG2]][4]
+//      CHECK:   %[[INSERT_ARG4:.+]] = llvm.insertvalue %[[PROCESSOR_ID]], %[[INSERT_ARG3]][5]
+//      CHECK:   llvm.store %[[INSERT_ARG4]], %[[PARAMSTRUCT_ALLOCA]]
+//      CHECK:   llvm.call @paramstruct_cconv_with_extra_fields(%[[PARAMSTRUCT_ALLOCA]])
+
+// -----
+
+module attributes {
+  hal.executable.target = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir]: Why is this using all the hal stuff? You should be able to put the target on just the function to get the target features.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir (hunk 2 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir]: Sorry, bad merge. I must have added those tests back as part of the conflict resolution. This PR is only adding the bitcast/interleave test.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**